### PR TITLE
Add file uploader support

### PR DESCRIPTION
**Pull request by FledgeXu**
# Goal
The purpose of this PR is to add an api to the webui that allows users to upload files.

# Abstract
This PR includes those changes as follows:
1. Add the `kiwixstorage` and `python-multipart` to dependencies.
2. Add the `upload`, `delete`, `update` and `get`, four apis.

### CODE REVIEW

**Comment by FledgeXu** [comment]
@rgaudin At the moment I haven't added S3 support, only local cache files. But I think the size of this PR is already too big, S3 support will be put in the next PR.

**Comment by FledgeXu** [comment]
Also while I was writing a new test, I found some errors in the previous test file, and I'll open a pr to fix those later as well.

**Comment by rgaudin** [comment]
OK ; S3 indeed is another feature, even though it will eventually happen right after upload

File: backend/api/routes/files.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+            path=str(location.resolve()),
+            type=content_type,
+            status="LOCAL",
+        )
+        project.files.append(new_file)
+        session.add(new_file)
+        session.flush()
+        session.refresh(new_file)
+        added_files.append(new_file)
+    return TypeAdapter(list[FileModel]).validate_python(added_files)
```
Comment (rgaudin) [inline backend/api/routes/files.py:103]: I think creation (POST) should return a _standard_ Success/Failure response with the rejection reason or the File ID

File: backend/api/routes/files.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+@router.get("/{file_id}", response_model=FileModel)
+async def get_file(project: File = Depends(validated_file)) -> FileModel:
+    """Get a specific file by its id."""
+    return FileModel.model_validate(project)
+
+
+@router.patch("/{file_id}", status_code=codes.NO_CONTENT)
+async def update_file(
+    file_request: FileRequest,
```
Comment (rgaudin) [inline backend/api/routes/files.py:122]:
```suggestion
    update_request: FileRequest,
```

**Review by rgaudin — CHANGES_REQUESTED** [review]
Thank you ;

- This is a multiple-files uploader while discussions had we'd have the UI manage (and track) multiple individual uploads. With your approach, you can't track individual files' progress and have to wait for all files' upload to get a feedback.

- There is no refusal management. We need to be able to reject some files.

- We need to discuss/settle the flows between the UI and the API first. Kind of design the API's I/O and implement it based on that.

- I wonder if we should prefix all routes with the project_id. All those endpoint are project specific. Maybe `/projects/23/files` is more appropriate. In this case, we should ensure that both the cookie project_id and path-based one are identical.

**Comment by rgaudin** [comment]
> > This is a multiple-files uploader while discussions had we'd have the UI manage (and track) multiple individual uploads. With your approach, you can't track individual files' progress and have to wait for all files' upload to get a feedback.
> 
> Maybe we need to let the files upload one by one. But I'm not sure how to implement the progress feature.

What I have in mind is that when reacting to the drag event, you gather a list of the dragged files ; each containing:
- filename (you get that from JS API)
- filesize (idem)
- a status that you maintain
- an xhr request that you create.
the xhr request emits events that informs you about the number of bytes transmitted so you can figure out the progress from filesize.

> > We need to discuss/settle the flows between the UI and the API first. Kind of design the API's I/O and implement it based on that.
> 
> I agree with that, but I don't have any particular ideas on that point, and I'm not sure where to start.

Well just listing the endpoints and what the response or responses for each could be. You can just add a comment here. Mostly useful for endpoints that can raise custom errors.

> > I wonder if we should prefix all routes with the project_id. All those endpoint are project specific. Maybe /projects/23/files is more appropriate. In this case, we should ensure that both the cookie project_id and path-based one are identical.
> 
> Yes, We can.

Great!

**Comment by FledgeXu** [comment]
>

> > > This is a multiple-files uploader while discussions had we'd have the UI manage (and track) multiple individual uploads. With your approach, you can't track individual files' progress and have to wait for all files' upload to get a feedback.
> > 
> > 
> > Maybe we need to let the files upload one by one. But I'm not sure how to implement the progress feature.
> 
> What I have in mind is that when reacting to the drag event, you gather a list of the dragged files ; each containing:
> 
> * filename (you get that from JS API)
> * filesize (idem)
> * a status that you maintain
> * an xhr request that you create.
>   the xhr request emits events that informs you about the number of bytes transmitted so you can figure out the progress from filesize.

Sorry, I don't have a particular understanding here. My current understanding is that the frontend sends data via XHR to the FastAPI backend, but does this have any impact on the backend, and do we keep the current API design? Or is it something else.

> > > We need to discuss/settle the flows between the UI and the API first. Kind of design the API's I/O and implement it based on that.
> > 
> > 
> > I agree with that, but I don't have any particular ideas on that point, and I'm not sure where to start.
> 
> Well just listing the endpoints and what the response or responses for each could be. You can just add a comment here. Mostly useful for endpoints that can raise custom errors.

Does it mean that I just need to add enough comments? Could you give me some examples?

File: backend/api/routes/files.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    """Upload new files"""
+    added_files = []
+    temp_file_location = BackendConf.cache_path.joinpath("files")
+    for file in uploaded_files:
+        binary_file = await file.read()
+
+        location = upload_file(temp_file_location, binary_file)
+
+        now = datetime.datetime.now(tz=datetime.UTC)
+        filename = file.filename if file.filename else ""
```
[the remark this answers is not in the document]
Comment (FledgeXu) [reply backend/api/routes/files.py:83]: I don't believe that uploading files will not have the names either. However, the type of the `file.filename` is `str | None`, I have to suppress pyright warnings.

File: backend/api/routes/files.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+@router.post("", response_model=list[FileModel], status_code=codes.CREATED)
+async def upload_files(
+    uploaded_files: list[UploadFile],
+    project: Project = Depends(validated_project),
+    session: Session = Depends(gen_session),
+) -> list[FileModel]:
+    """Upload new files"""
+    added_files = []
+    temp_file_location = BackendConf.cache_path.joinpath("files")
```
[the remark this answers is not in the document]
Comment (FledgeXu) [reply backend/api/routes/files.py:76]: For now We don't have other things beside files. I will delete it later.

File: backend/api/routes/files.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rgaudin) [reply backend/api/routes/files.py:83]: A browser will most likely always supply one but a random API user could just make a request without it.

[the remark this answers is not in the document]
Comment (FledgeXu) [reply backend/api/routes/files.py:83]: That being the case if there is no filename then we should reject the request

File: backend/api/routes/files.py [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+
+    Raises:
+        HTTPException: If the filename is invalid, the file is empty.
+    """
+    filename = upload_file.filename
+    size = upload_file.size
+
+    if not filename:
+        raise HTTPException(
+            status_code=codes.NOT_ACCEPTABLE, detail="Filename is invalid."
```
Comment (FledgeXu) [inline backend/api/routes/files.py:115]: I'm not sure how to test this part. `TestClient` doesn't support the `None` as the filename.

File: backend/api/routes/files.py [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+            detail="Uploaded files exceeded quota",
+        )
+
+
+@router.post("/{project_id}/files", status_code=codes.CREATED)
+async def create_file(
+    uploaded_file: UploadFile,
+    project: Project = Depends(validated_project),
+    session: Session = Depends(gen_session),
+) -> FileModel:
```
Comment (FledgeXu) [inline backend/api/routes/files.py:142]: The api can only upload one file at a time, making it easier to track the progress of file uploads.

**Review by FledgeXu — COMMENTED** [review]
I encountered a problem when adding tests, the `TestClient` provided by FastAPI doesn't allow the use of None as a filename, so I couldn't cover the Filename as None part of the code.

File: backend/api/routes/files.py [hunk 7]
```diff
@@ -52,74 +61,145 @@ def validated_file(
...
+        # TODO: Using S3 to save file.
+        path=str(location.resolve()),
+        type=mimetype,
+        status=FileSaveLocation.LOCAL.value,
+    )
+    project.files.append(new_file)
+    session.add(new_file)
+    session.flush()
+    session.refresh(new_file)
+    return FileModel.model_validate(new_file)
```
Comment (rgaudin) [inline backend/api/routes/files.py:183]: Code should be `201`

**Review by rgaudin — CHANGES_REQUESTED** [review]
One thing we can add in a future PR is disk space management for local storage. We've discussed it in the past ; you can open a ticket.
Basically we'd want to define a local disk size and make sure we stop writing to disk once we hit it. How this is going to be reported to users is another discussion.

keep in mind disk writes may fail : disk exhausted, permissions being incorrect, physical disk issue.

To track files on local filesystem, I suggest we use `{project_id}-{hash}` for local filename. Then we don't have to keep a reference to the local filename. Calculating hash first will help us skip writing if we already have the same file (in the project!) and will save us from uploading it as well.

File: backend/tests/routes/test_files.py [hunk 8]
```diff
@@ -3,54 +3,74 @@
...
 def test_get_all_files_correct_data(logged_in_client, project_id, file_id):
     params = {"project_id": project_id}
-    response = logged_in_client.get(f"{API_VERSION_PREFIX}/files", params=params)
+    response = logged_in_client.get(
+        f"{API_VERSION_PREFIX}/projects/{project_id}/files", params=params
+    )
     json_result = response.json()
     assert response.status_code == codes.OK
     assert json_result is not None
     assert len(json_result) == 1
```
[the remark this answers is not in the document]
Comment (FledgeXu) [reply backend/tests/routes/test_files.py:57]: Since this API returns information about all the Files of a project, it is correct that this is a list.

[the remark this answers is not in the document]
Comment (rgaudin) [reply backend/tests/routes/test_files.py:57]: My bad, it is indeed a list request

File: backend/api/routes/files.py [hunk 9]
```diff
@@ -52,74 +61,145 @@ def validated_file(
...
+    """Saves a binary file to a specific location and returns its path."""
     if not location.exists():
         os.makedirs(location, exist_ok=True)
     file_location = Path(tempfile.NamedTemporaryFile(dir=location, delete=False).name)
-    with open(file_location, "wb+") as file_object:
-        file_object.write(file)
+    with open(file_location, "wb") as file_object:
+        for chunk in read_file_in_chunks(file):
+            file_object.write(chunk)
     return file_location
```
[the remark this answers is not in the document]
Comment (FledgeXu) [reply backend/api/routes/files.py:94]: `file` is already `seek(0)` in `read_file_in_chunks`, no need to seek here.

[the remark this answers is not in the document]
Comment (rgaudin) [reply backend/api/routes/files.py:94]: This is a bad pattern. You don't want to rely on another function restoring state for stuff you do here. What happens when you refactor and change that other function? Weird bugs. It's a zero cost operation

[the remark this answers is not in the document]
Comment (FledgeXu) [reply backend/api/routes/files.py:94]: Ok, I will change it then.

File: backend/api/routes/files.py (hunk 7 above)
Comment (FledgeXu) [reply backend/api/routes/files.py:183]: `status_code` is `codes.CREATED(201)` now. I will change it to `http` lib.

File: backend/api/routes/files.py [hunk 10]
```diff
@@ -52,74 +61,145 @@ def validated_file(
...
+        filesize=size,
+        title=filename,
+        authors=None,
+        description=None,
+        uploaded_on=now,
+        hash=generate_file_hash(uploaded_file.file),
+        # TODO: Using S3 to save file.
+        path=str(location.resolve()),
+        type=mimetype,
+        status=FileSaveLocation.LOCAL.value,
```
[the remark this answers is not in the document]
Comment (rgaudin) [reply backend/api/routes/files.py:177]: Right

File: backend/tests/routes/test_files.py [hunk 11]
```diff
@@ -134,16 +166,24 @@ def test_update_file_wrong_data(logged_in_client, project_id, non_existent_file_
     }
     params = {"project_id": project_id}
     response = logged_in_client.patch(
-        f"{API_VERSION_PREFIX}/files/{non_existent_file_id}", json=data, params=params
+        f"{API_VERSION_PREFIX}/projects/{project_id}/files/{non_existent_file_id}",
+        json=data,
+        params=params,
```
[the remark this answers is not in the document]
Comment (FledgeXu) [reply backend/tests/routes/test_files.py:171]: `test_update_file_wrong_data` is the test what you asked.

`test_update_file_correct_data`: correct auth and correct file_id and correct payload.
`test_update_file_wrong_data`: correct auth and correct file_id but incorrect payload.
`test_update_file_wrong_authorization`: incorrect auth.

[the remark this answers is not in the document]
Comment (FledgeXu) [reply backend/tests/routes/test_files.py:171]: I put the wrong id here. I'll fix it.

[the remark this answers is not in the document]
Comment (FledgeXu) [reply backend/tests/routes/test_files.py:217]: It's true that it's not quite right, the name and content of this function is not quite right, I'll add another one.

File: backend/api/routes/files.py [hunk 12]
```diff
@@ -83,15 +84,25 @@ def generate_file_hash(file: BinaryIO) -> str:
     return hasher.hexdigest()
 
 
-def upload_file(location: Path, file: BinaryIO) -> Path:
+def save_file(file_name: str, file: BinaryIO) -> Path:
```
Comment (rgaudin) [inline backend/api/routes/files.py:87]: I think it makes more sense to have the file (main subject) be the first arg
```suggestion
def save_file(file: BinaryIO, file_name: str) -> Path:
```

File: backend/api/routes/files.py [hunk 13]
```diff
@@ -83,15 +84,25 @@ def generate_file_hash(file: BinaryIO) -> str:
...
-            file_object.write(chunk)
-    return file_location
+    if not BackendConf.transient_storage_path.exists():
+        os.makedirs(BackendConf.transient_storage_path, exist_ok=True)
+    fpath = Path(BackendConf.transient_storage_path).joinpath(file_name)
+    if not fpath.is_file():
+        with open(fpath, "wb") as file_object:
+            for chunk in read_file_in_chunks(file):
+                file_object.write(chunk)
+            file.seek(0)
```
Comment (rgaudin) [inline backend/api/routes/files.py:96]: Not needed here. You are not reading directly from `file` in this function.

File: backend/api/routes/files.py [hunk 14]
```diff
@@ -170,12 +193,13 @@ async def create_file(
         authors=None,
         description=None,
         uploaded_on=now,
-        hash=generate_file_hash(uploaded_file.file),
+        hash=file_hash,
         # TODO: Using S3 to save file.
-        path=str(location.resolve()),
+        path=str(fpath.resolve()),
```
Comment (rgaudin) [inline backend/api/routes/files.py:198]: no need for the resolve()

File: backend/api/routes/files.py [hunk 15]
```diff
@@ -217,11 +241,19 @@ async def update_file(
...
 ):
     """Delete a specific file by its id."""
-    file_location = Path(file.path)
-    file_location.unlink(missing_ok=True)
+    stmt = (
+        select(func.count())
+        .select_from(File)
+        .filter_by(project_id=file.project_id)
+        .filter_by(hash=file.hash)
+    )
```
Comment (rgaudin) [inline backend/api/routes/files.py:254]: This should filter local files (using status) and then reuse the hash to find the local filename. The `path` hack is meant to disapear.

**Review by rgaudin — CHANGES_REQUESTED** [review]

File: backend/api/routes/files.py (hunk 15 above)
Comment (FledgeXu) [reply backend/api/routes/files.py:254]: I don't quite understand here. How could I avoid the `path` hacker here.

Comment (rgaudin) [reply backend/api/routes/files.py:254]: The local filepath is constructed by concatenating the transient storage path with the hash. Since you have the hash, you can just rebuild the local file path

Comment (FledgeXu) [reply backend/api/routes/files.py:254]: I'm still not quite sure what the `path` hack refers to here and why we should rebuild the local file path from the hash.
Is this right?
```python
    stmt = (
        select(func.count())
        .select_from(File)
        .filter_by(project_id=file.project_id)
        .filter_by(hash=file.hash)
        .filter_by(status=FileSaveLocation.LOCAL.value)
    )
    number_of_duplicate_files = session.scalars(stmt).one()
    if number_of_duplicate_files == 1:
        file_location = Path(f"{file.project_id}-{file.hash}")
        file_location.unlink(missing_ok=True)
    session.delete(file)
```

File: backend/api/routes/files.py [hunk 16]
```diff
@@ -84,16 +83,14 @@ def generate_file_hash(file: BinaryIO) -> str:
     return hasher.hexdigest()
 
 
-def save_file(file_name: str, file: BinaryIO) -> Path:
+def save_file(file: BinaryIO, file_hash: str, project_id: UUID) -> Path:
     """Saves a binary file to a specific location and returns its path."""
-    if not BackendConf.transient_storage_path.exists():
-        os.makedirs(BackendConf.transient_storage_path, exist_ok=True)
-    fpath = Path(BackendConf.transient_storage_path).joinpath(file_name)
+    BackendConf.transient_storage_path.mkdir(exist_ok=True)
```
Comment (rgaudin) [inline backend/api/routes/files.py:88]: Given this is a _constant_, it shouldn't be any function using it's responsibility to ensure folder exists. This line should be in the BackendConf

File: backend/api/routes/files.py [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+
+from api.constants import BackendConf
+from api.database import gen_session
+from api.database.models import File, Project
+from api.routes import validated_project
+
+router = APIRouter(prefix="/files")
+
+
+class FileRequest(BaseModel):
```
[the remark this answers is not in the document]
Comment (rgaudin) [reply backend/api/routes/files.py:24]: Please pay attention, I've already explained this.

This is not a `FileRequest`, it's a Metadata Update request for a file ; or an Update Request for a File. Name it accordingly.

[the remark this answers is not in the document]
Comment (rgaudin) [reply backend/api/routes/files.py:24]: Now please also rename the `FileSaveLocation` into `FileStatus` as that's what it is.

**Review by rgaudin — CHANGES_REQUESTED** [review]

**Review by rgaudin — APPROVED** [review]
Thank you ; I made the constants change myself
